### PR TITLE
Angular: Add angular 14 sandbox template

### PR DESCRIPTION
**Pull request by tmeasday**
## What I did

Add template for angular 14 and template components for angular 🎉 .

## Caveats

The angular template does not work out-of-the-box with `yarn bootstrap --prep` (It works fine with `--core`). In particular:

1. If you `--no-link` (i.e. verdaccio), it has many TS errors in our code that you can remove by adding `"skipLibCheck": true` to the generated `.storybook/tsconfig.json`.
2. If you `--link`, it seems the `skipLibCheck` doesn't work because it doesn't respect symlinks and thus thinks our code isn't in `node_modules` (at least I think that's what going on.

We are going to follow up the above -- some options are:
 - Fix the TS problems
 - Loosen the TS setting for the angular sandbox.

### CODE REVIEW

**Comment by ndelangen** [comment]
@tmeasday It's unfortunate that they are marked as `Error` because otherwise we could have added them to a ignore list with `ignoreWarnings = https://webpack.js.org/configuration`

Perhaps something like this is an option:
```ts
module.exports = {
  ...
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: [
          {
            loader: 'ts-loader',
            options: {
              transpileOnly: true
            }
          }
        ]
      }
    ]
  }
}
```
source: https://stackoverflow.com/questions/60278796/how-to-ignore-typescript-errors-when-compile-by-webpack-production-mode

**Comment by tmeasday** [comment]
@ndelangen @kroeder (and anyone else that can help!) - [here are the issues](https://www.chromatic.com/build?appId=6322ce6af69825592bbb28fc&number=6) I am seeing on this PR:

1. Our pattern of passing args via a `decorator` function isn't working: 

https://github.com/storybookjs/storybook/blob/1e39b02c1001d2e527d74558f6298267e3d6ec38/code/lib/store/template/stories/argTypes.stories.ts#L8-L12

I don't know angular well enough to say for sure, but I'm reasonably sure this isn't resulting in `args.object` being set on the rendered story.

2. Similarly, in this story, the button renders as "Button" not "Clicked X times"

https://github.com/storybookjs/storybook/blob/1e39b02c1001d2e527d74558f6298267e3d6ec38/code/lib/store/template/stories/hooks.stories.ts#L15-L23

https://www.chromatic.com/test?appId=6322ce6af69825592bbb28fc&id=633e5fc2f266eeeadf703d4f: 

image

3. These stories [with a callback](https://github.com/storybookjs/storybook/blob/1e39b02c1001d2e527d74558f6298267e3d6ec38/code/addons/interactions/template/stories/basics.stories.ts#L41) are failing: https://6322ce6af69825592bbb28fc-uvegsxykpm.chromatic.com/?path=/story/addons-interactions-basics--callback

The component definitely emits `onSuccess` but it doesn't seem to trigger the action. I copied what the button story does but it doesn't seem to be working. Any ideas?

**Comment by kroeder** [comment]
For the callback examples, I'm not even sure what to expect. Let's take a look at hooks.stories.ts

It uses "ButtonComponent" I assume? The story uses globalThis.Components.Button and I don't know where this is defined.

**TL;DR** **"onClick" gets passed to the component, but "onClick" is an output, not an input. I'm confused - why should this work? Has it worked before? Chances are, I misunderstand the code.**

ButtonComponent has a property

```
  @Output()
  onClick = new EventEmitter<Event>();
```

In the story, we pass in "onClick" with the same name as the output 

```
      return story({
        args: {
          children: `Clicked ${count} times`,
          onClick: () => {
            setCount(count + 1);
          },
        },
      });
```

But onClick is not an input of a component but an output. Outputs push updates to the outside world 
https://angular.io/guide/inputs-outputs#sending-data-to-a-parent-component

A quick example

```html
<storybook-button [label]="'this is an input; passed to the component'" (buttonClicked)="thisIsAnOutput($event)"></storybook-button>
```

To trigger an output in a component, you need to emit a value

```typescript
class StorybookButton {
  
  @Output() buttonClicked = new EventEmitter<string>();

  foo() {
    this.buttonClicked.emit("Hello world!");
  }
}
```

The outside world can now subscribe to the output in a template like I did before:

```html
<storybook-button (buttonClicked)="thisIsAnOutput($event)"></storybook-button>
```

Or reference the component in TypeScript and subscribe to the output

```typescript
// outside world
storybookButtonRef.buttonClicked.subscribe(value => console.log(value));
```

**Comment by tmeasday** [comment]
Hey @kroeder thanks for looking at it.

> It uses "ButtonComponent" I assume? The story uses globalThis.Components.Button and I don't know where this is defined.

In these "generalised stories" for the sandbox (i.e. stories that work in all frameworks), we need to define the component in a slightly obtuse way. Here is the point where it gets setup: https://github.com/storybookjs/storybook/blob/cc723f8cd67a440f262e4e5668b9e2d7e4cdef68/code/frameworks/angular/template/components/index.js#L8

> The outside world can now subscribe to the output in a template like I did before...

So I was trying to imitate the CLI template's [`button.component.ts`](https://github.com/storybookjs/storybook/blob/cc723f8cd67a440f262e4e5668b9e2d7e4cdef68/code/lib/cli/rendererAssets/angular/button.component.ts#L46) and [`Button.stories.ts`](https://github.com/storybookjs/storybook/blob/cc723f8cd67a440f262e4e5668b9e2d7e4cdef68/code/lib/cli/rendererAssets/angular/Button.stories.ts#L1) in which the action (which I assume is bound to the `onClick` arg) does trigger when the button is clicked and [the `onClick` event is triggered](https://github.com/storybookjs/storybook/blob/cc723f8cd67a440f262e4e5668b9e2d7e4cdef68/code/lib/cli/rendererAssets/angular/button.component.ts#L7).

What am I getting wrong here? Is it not possible to pass a functional arg to a Angular story and have it trigger when an output is emitted?

> Our pattern of passing args via a decorator function isn't working

This one seems a little more complex to me. Are the decorators I linked to (that end up calling `storyFn({ args: ... })`) not going to work in Angular? 

That's a technique that we've leaned on quite a bit in these generalised stories so I'd be super keen to hear if it is an issue.

**Comment by kroeder** [comment]
I played around with templates and encountered some API changes I was not aware of but I think I'm at a point where I understand the problem(s). 

I currently investigate the argTypes story issue and get back to you. If I get stuck as well, we can also try to pair on this by end of next week or the week after

**Comment by kroeder** [comment]
@tmeasday how comes I cannot see any updates in the storyFn result? Is that a bug or does this work in a different way?

```typescript
    (storyFn: PartialStoryFn, context: StoryContext) => {
      const fn = storyFn({args: {object: context.argTypes}});
      console.log(fn);
      return fn;
    }],
```

I tried to set a value on a control and expected to have this value at least in the "props" field in `console.log(fn)` but what I get is this:

```
{
    "props": {},
    "template": "<storybook-pre></storybook-pre>",
    "userDefinedTemplate": false
}
```

**Comment by tmeasday** [comment]
> I tried to set a value on a control and expected to have this value at least in the "props" field in console.log(fn) but what I get is this:

@kroeder this is getting into a part of the code base I don't know so well and was hoping you might know ;)

So the `storyFn` in a decorator is the decorated (with all the previous decorators) version of the so-called `originalStoryFn`. In a CSF3 file like this one, the default implementation of a story function will be angular's `render` function[1]:

https://github.com/storybookjs/storybook/blob/07a0f39cbc28ff16c75b30ebcde2fa35403be069/code/frameworks/angular/src/client/render.ts#L11

(Apparently I wrote that line although don't really remember it).

I guess it seems like Angular's `applyDecorators` implementation is somehow losing track of the args. It could have something to do with `cleanArgsDecorator` -- do you know that's for?

[1] there is some talk of renaming this to the `template` function, btw.

**Comment by tmeasday** [comment]
> I guess it seems like Angular's applyDecorators implementation is somehow losing track of the args. It could have something to do with [cleanArgsDecorator](https://github.com/storybookjs/storybook/blob/07a0f39cbc28ff16c75b30ebcde2fa35403be069/code/frameworks/angular/src/client/decorateStory.ts#L53-L72) -- do you know that's for?

I confirmed that removing the `cleanArgsDecorator` fixed the problem. Going to do some git blame spelunking to figure out what it's for.

**Comment by tmeasday** [comment]
Discussed with @shilman and we worked out the reason was that compodoc wasn't running properly due to symlinks (an already known issue). Angular has some guardrails (which we could discuss separately) to check that unknown `args` are droped.

**Review by shilman — APPROVED** [review]
Great job!! 💯

File: code/frameworks/angular/template/stories/basics/component-with-provider/di.component.stories.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
+import type { Args } from '@storybook/angular';
 import { DiComponent } from './di.component';
 
 export default {
-  title: 'Basics / Component / With Provider',
+  // title: 'Basics / Component / With Provider',
```
Comment (ThibaudAV) [inline code/frameworks/angular/template/stories/basics/component-with-provider/di.component.stories.ts:5]: the title is deactivated in order to take the one of the file ?

**Comment by kroeder** [comment]
Creating a template does not work anymore

- `yarn sandbox --template angular-cli/default-ts` seems to be removed 
- `yarn task --task sandbox --template angular-cli/default-ts` 

image

then selecting the first option `Create the sandbox from a template (sandbox)`
results in an error


```
👷 Bootstrapping Template
> node /Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js repro-next angular-cli/default-ts --output /Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts --branch next
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                     │
│   🔎 You filtered out all templates. 🔍                                                                                             │
│                                                                                                                                     │
│   After filtering all the templates with "angular-cli/default-ts", we found no results. Please try again with a different filter.   │
│                                                                                                                                     │
│   Available templates:                                                                                                              │
│   - cra/default-js                                                                                                                  │
│   - cra/default-ts                                                                                                                  │
│   - react-webpack/18-ts                                                                                                             │
│   - react-webpack/17-ts                                                                                                             │
│   - react-vite/default-js                                                                                                           │
│   - react-vite/default-ts                                                                                                           │
│   - vue2-vite/2.7-js                                                                                                                │
│   - vue3-vite/default-js                                                                                                            │
│   - vue3-vite/default-ts                                                                                                            │
│   - svelte-vite/default-js                                                                                                          │
│   - svelte-vite/default-ts                                                                                                          │
│   - lit-vite/default-js                                                                                                             │
│   - lit-vite/default-ts                                                                                                             │
│   - vue-cli/default-js                                                                                                              │
│   - vue-cli/vue2-default-js                                                                                                         │
│                                                                                                                                     │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
An error occurred while executing: `node /Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js repro-next angular-cli/default-ts --output /Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts --branch next`
🚨 Bootstrapping Template failed

Error running task sandbox:

Error: Command failed with exit code 1: node /Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js repro-next angular-cli/default-ts --output /Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts --branch next
    at makeError (/Users/kroeder/WebstormProjects/storybook/scripts/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/Users/kroeder/WebstormProjects/storybook/scripts/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async exec (/Users/kroeder/WebstormProjects/storybook/scripts/utils/exec.ts:43:7)
    at async executeCLIStep (/Users/kroeder/WebstormProjects/storybook/scripts/utils/cli-step.ts:78:3)
    at async create (/Users/kroeder/WebstormProjects/storybook/scripts/tasks/sandbox-parts.ts:62:5)
    at async Object.run (/Users/kroeder/WebstormProjects/storybook/scripts/tasks/sandbox.ts:20:5)
    at async runTask (/Users/kroeder/WebstormProjects/storybook/scripts/task.ts:271:5)
    at async run (/Users/kroeder/WebstormProjects/storybook/scripts/task.ts:401:9) {
  shortMessage: 'Command failed with exit code 1: node /Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js repro-next angular-cli/default-ts --output /Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts --branch next',
  command: 'node /Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js repro-next angular-cli/default-ts --output /Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts --branch next',
  escapedCommand: 'node "/Users/kroeder/WebstormProjects/storybook/code/lib/cli/bin/index.js" repro-next "angular-cli/default-ts" --output "/Users/kroeder/WebstormProjects/storybook/sandbox/angular-cli-default-ts" --branch next',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: undefined,
  stderr: '',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

File: code/frameworks/angular/template/stories/basics/component-with-provider/di.component.stories.ts (hunk 1 above)
Comment (tmeasday) [reply code/frameworks/angular/template/stories/basics/component-with-provider/di.component.stories.ts:5]: Yeah, one thing we wanted to do was to rely on autotitle so it was easier to map stories back to the files they came from. I for one have often struggled to figure out which file a failing story is from due to titles. I left the names in there commented out in case someone wanted to rename the files to better match the old titles.

**Comment by tmeasday** [comment]
Ughh, bad merge sorry.

**Review by ndelangen — APPROVED** [review]
